# AdInsight AI

# Module 01 – Campaign Analytics

## Objective

Analyze campaign performance to identify the most valuable traffic sources, marketing channels, devices, countries, and campaigns using Google Analytics data.

---

## Business Questions

1. Which marketing channels generate the highest traffic?
2. Which traffic sources generate the highest sessions?
3. Which countries generate the highest traffic?
4. Which devices dominate website usage?
5. Which browsers are most commonly used?
6. Which campaigns perform best?
7. Which channels generate the highest revenue?
8. What are the daily traffic trends?

In [8]:
import sys
from pathlib import Path

# Get project root
PROJECT_ROOT = Path.cwd().parent

# Add project root to Python path
if str(PROJECT_ROOT) not in sys.path:
    sys.path.append(str(PROJECT_ROOT))

In [9]:
# =====================================================
# Import Libraries
# =====================================================

import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

from config.config import MASTER_DATASET

In [10]:
# =====================================================
# Load Dataset
# =====================================================

df = pd.read_parquet(MASTER_DATASET)

print("=" * 60)
print("Dataset Loaded Successfully")
print("=" * 60)

print(f"Shape : {df.shape}")

df.head()

Dataset Loaded Successfully
Shape : (100000, 18)


,fullVisitorId,visitId,visit_date,channelGrouping,source,medium,campaign,deviceCategory,browser,operatingSystem,country,pageviews,hits,timeOnSite,transactions,transactionRevenue,newVisits,bounces
0,4598890394865817235,1486932897,2017-02-12,Direct,(direct),(none),(not set),mobile,Android Webview,Android,Netherlands,1,1,<NA>,<NA>,<NA>,1,1
1,085297959192196865,1486906252,2017-02-12,Organic Search,google,organic,(not set),mobile,Chrome,Android,United Kingdom,1,1,<NA>,<NA>,<NA>,<NA>,1
2,2211677311320024670,1486956669,2017-02-12,Direct,(direct),(none),(not set),mobile,Chrome,Android,Argentina,1,1,<NA>,<NA>,<NA>,<NA>,1
3,8969779493563769417,1486922640,2017-02-12,Organic Search,google,organic,(not set),mobile,Safari,iOS,United States,1,1,<NA>,<NA>,<NA>,<NA>,1
4,9271977665405126573,1486922683,2017-02-12,Organic Search,google,organic,(not set),desktop,Safari,Macintosh,United States,1,1,<NA>,<NA>,<NA>,<NA>,1


In [11]:
# =====================================================
# Dataset Information
# =====================================================

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 18 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   fullVisitorId       100000 non-null  str           
 1   visitId             100000 non-null  Int64         
 2   visit_date          100000 non-null  datetime64[us]
 3   channelGrouping     100000 non-null  str           
 4   source              100000 non-null  str           
 5   medium              100000 non-null  str           
 6   campaign            100000 non-null  str           
 7   deviceCategory      100000 non-null  str           
 8   browser             100000 non-null  str           
 9   operatingSystem     100000 non-null  str           
 10  country             100000 non-null  str           
 11  pageviews           99987 non-null   Int64         
 12  hits                100000 non-null  Int64         
 13  timeOnSite          48192 non-null   Int6

In [13]:
# =====================================================
# Executive KPI Dashboard
# =====================================================

kpi_df = pd.DataFrame({
    "Metric": [
        "Total Sessions",
        "Unique Users",
        "Transactions",
        "Revenue",
        "Average Pageviews",
        "Average Time on Site",
        "Bounce Rate"
    ],
    "Value": [
        f"{total_sessions:,}",
        f"{unique_users:,}",
        f"{transactions:,}",
        f"${revenue:,.2f}",
        f"{avg_pageviews:.2f}",
        f"{avg_time:.2f} sec",
        f"{bounce_rate:.2f}%"
    ]
})

kpi_df

,Metric,Value
0,Total Sessions,"100,000"
1,Unique Users,"89,989"
2,Transactions,"1,322"
3,Revenue,"$181,588.74"
4,Average Pageviews,3.73
5,Average Time on Site,129.19 sec
6,Bounce Rate,51.69%


# Executive KPI Interpretation

## Business Objective

The Executive KPI Dashboard provides a high-level overview of website performance and user engagement. These metrics help marketing teams and business stakeholders quickly assess overall website health before performing detailed campaign analysis.

---

## Key Observations

### Website Traffic
- The website recorded **100,000 total sessions** from **89,989 unique users**.
- This indicates that a portion of users returned to the website more than once, suggesting repeat visitor engagement.

### User Engagement
- Visitors viewed an average of **3.73 pages per session**.
- The average time spent on the website was **129.19 seconds** (approximately **2.15 minutes**), indicating moderate user engagement.

### Conversion Performance
- The website generated **1,322 transactions**.
- Total revenue generated during the selected period was **$181,588.74**.

### Bounce Behaviour
- The bounce rate is **51.69%**, meaning that approximately half of all sessions ended after viewing only one page.
- This suggests there is room to improve landing page quality and user engagement.

---

## Business Recommendations

Based on the Executive KPI Dashboard:

- Improve landing page experience to reduce bounce rate.
- Increase user engagement by optimizing website navigation and content.
- Analyze high-performing marketing channels to understand which campaigns generate the highest revenue.
- Focus future analysis on identifying the traffic sources and campaigns responsible for the highest conversions and revenue.

---

## Next Analysis

The next section will analyze **Marketing Channels** to identify which acquisition channels drive the highest traffic and business value.

# Section 2 – Marketing Channel Analysis

## Business Question

Which marketing channels generate the highest number of website sessions?

### Why is this important?

Understanding traffic acquisition channels helps businesses identify where users are coming from and where marketing efforts should be focused.

Examples of marketing channels include:

- Organic Search
- Direct
- Referral
- Social
- Paid Search
- Display
- Email

In [14]:
# =====================================================
# Traffic by Marketing Channel
# =====================================================

traffic_channel = (
    df.groupby("channelGrouping")
      .size()
      .reset_index(name="Sessions")
      .sort_values("Sessions", ascending=False)
)

# Calculate percentage contribution
traffic_channel["Percentage"] = (
    traffic_channel["Sessions"] /
    traffic_channel["Sessions"].sum()
) * 100

traffic_channel["Percentage"] = traffic_channel["Percentage"].round(2)

traffic_channel

,channelGrouping,Sessions,Percentage
4,Organic Search,42337,42.34
7,Social,25259,25.26
2,Direct,16035,16.04
6,Referral,11335,11.34
5,Paid Search,2471,2.47
1,Affiliates,1898,1.90
3,Display,648,0.65
0,(Other),17,0.02


In [17]:
# =====================================================
# Professional Executive Bar Chart
# =====================================================

fig = px.bar(
    traffic_channel,
    x="channelGrouping",
    y="Sessions",
    text="Sessions",
    title="<b>Traffic Acquisition by Marketing Channel</b>",
    color_discrete_sequence=["#4285F4"]   # Google Blue
)

fig.update_traces(
    texttemplate="%{text:,.0f}",
    textposition="outside",
    marker_line_width=0,
    marker_opacity=0.95,
    hovertemplate=
    "<b>%{x}</b><br>" +
    "Sessions: %{y:,}<extra></extra>"
)

fig.update_layout(

    template="simple_white",

    title=dict(
        x=0.5,
        font=dict(size=24)
    ),

    font=dict(
        family="Arial",
        size=14,
        color="#202124"
    ),

    plot_bgcolor="white",
    paper_bgcolor="white",

    xaxis=dict(
        title="Marketing Channel",
        showgrid=False,
        tickangle=-20
    ),

    yaxis=dict(
        title="Sessions",
        gridcolor="#E8EAED",
        zeroline=False
    ),

    showlegend=False,

    margin=dict(
        t=80,
        l=70,
        r=30,
        b=70
    ),

    height=550
)

fig.show()

In [18]:
fig = px.pie(

    traffic_channel,

    names="channelGrouping",

    values="Sessions",

    hole=0.60,

    title="<b>Traffic Distribution by Marketing Channel</b>",

    color_discrete_sequence=[
        "#4285F4",
        "#34A853",
        "#FBBC05",
        "#EA4335",
        "#7E57C2",
        "#00ACC1",
        "#F4511E",
        "#9AA0A6"
    ]
)

fig.update_traces(

    textposition="inside",

    textinfo="percent",

    insidetextfont=dict(size=14),

    hovertemplate=
    "<b>%{label}</b><br>" +
    "Sessions: %{value:,}<br>" +
    "Share: %{percent}<extra></extra>"
)

fig.update_layout(

    template="simple_white",

    title=dict(
        x=0.5,
        font=dict(size=24)
    ),

    font=dict(
        family="Arial",
        size=14,
        color="#202124"
    ),

    legend=dict(
        orientation="v",
        x=1.02,
        y=0.5
    ),

    margin=dict(
        t=80,
        l=40,
        r=40,
        b=40
    ),

    height=600
)

fig.show()

# Section 3 – Traffic Source Analysis

## Business Question

Which traffic sources generate the highest number of website sessions?

### Why is this important?

Traffic source analysis helps businesses understand where visitors are coming from.

Instead of looking at broad marketing channels, this analysis identifies the exact websites or search engines responsible for bringing users to the platform.

Examples include:

- Google
- YouTube
- Facebook
- Bing
- Yahoo
- Direct traffic

In [19]:
# =====================================================
# Top Traffic Sources
# =====================================================

traffic_source = (
    df.groupby("source")
      .size()
      .reset_index(name="Sessions")
      .sort_values("Sessions", ascending=False)
      .head(10)
)

traffic_source["Percentage"] = (
    traffic_source["Sessions"] /
    traffic_source["Sessions"].sum()
    * 100
).round(2)

traffic_source

,source,Sessions,Percentage
0,(direct),37669,38.76
41,google,29832,30.69
142,youtube.com,23973,24.67
3,Partners,1899,1.95
5,analytics.google.com,1841,1.89
26,dfa,595,0.61
50,google.com,517,0.53
11,baidu,346,0.36
125,sites.google.com,296,0.30
114,reddit.com,221,0.23


In [38]:
# =====================================================
# Top 10 Website Traffic Sources
# Executive Dashboard Style
# =====================================================

import plotly.graph_objects as go

# -----------------------------------------------------
# Prepare Data
# -----------------------------------------------------

traffic_source = (
    df.groupby("source")
      .size()
      .reset_index(name="Sessions")
      .sort_values("Sessions", ascending=False)
      .head(10)
      .reset_index(drop=True)
)

# Calculate Traffic Share (%)
traffic_source["Share"] = (
    traffic_source["Sessions"] /
    traffic_source["Sessions"].sum() * 100
).round(1)

# Rank
traffic_source["Rank"] = (
    traffic_source.index + 1
)

# Labels
traffic_source["Label"] = (
    traffic_source["Sessions"]
    .map("{:,.0f}".format)
    + " ("
    + traffic_source["Share"].astype(str)
    + "%)"
)

# -----------------------------------------------------
# Google Material Color Palette
# -----------------------------------------------------

colors = [
    "#1967D2",  # Google Blue (Top Source)
    "#669DF6",
    "#8AB4F8",
    "#D2E3FC",
    "#DADCE0",
    "#DADCE0",
    "#DADCE0",
    "#DADCE0",
    "#DADCE0",
    "#DADCE0"
]

# -----------------------------------------------------
# Create Figure
# -----------------------------------------------------

fig = go.Figure()

fig.add_trace(

    go.Bar(

        x=traffic_source["Sessions"],

        y=traffic_source["source"],

        orientation="h",

        marker=dict(

            color=colors,

            line=dict(
                color="white",
                width=1
            )

        ),

        text=traffic_source["Label"],

        textposition="auto",

        textfont=dict(
            family="Arial",
            size=13
        ),

        cliponaxis=False,

        customdata=traffic_source[
            ["Share", "Rank"]
        ],

        hovertemplate=
        "<b>Traffic Source</b><br>"
        "%{y}<br><br>"
        "<b>Sessions</b><br>"
        "%{x:,}<br><br>"
        "<b>Traffic Share</b><br>"
        "%{customdata[0]}%<br><br>"
        "<b>Rank</b><br>"
        "#%{customdata[1]}"
        "<extra></extra>"
    )

)

# -----------------------------------------------------
# Layout
# -----------------------------------------------------

fig.update_layout(

    template="simple_white",

    title=dict(

        text="<b>Top 10 Website Traffic Sources</b>"
             "<br><sup>Google Analytics Sample Dataset • Top 10 Sources by Sessions</sup>",

        x=0.5,

        font=dict(
            size=28,
            family="Arial"
        )

    ),

    font=dict(

        family="Arial",

        size=16,

        color="#202124"

    ),

    plot_bgcolor="white",

    paper_bgcolor="white",

    xaxis=dict(

        title="",

        showgrid=True,

        gridcolor="#F1F3F4",

        zeroline=False,

        tickformat=",",

        range=[0, 40000]

    ),

    yaxis=dict(

        title=""

    ),

    margin=dict(

        l=130,

        r=60,

        t=130,

        b=60

    ),

    height=650,

    showlegend=False,

    hoverlabel=dict(

        bgcolor="white",

        bordercolor="#DADCE0",

        font_size=13,

        font_family="Arial"

    )

)
fig.update_traces(
    width=0.75
)
# -----------------------------------------------------
# Display Chart
# -----------------------------------------------------

fig.show()

# 🌍 Section 4 – Geographic Analysis

## Business Question

Which countries generate the highest website traffic and revenue?

### Why is this important?

Geographic analysis helps businesses understand where their customers are located. These insights support regional marketing strategies, localization, international expansion, and advertising budget allocation.

This section analyzes:

- Top countries by website sessions
- Top countries by unique users
- Revenue generated by each country
- Geographic distribution of website traffic

In [39]:
# =====================================================
# Geographic Performance
# =====================================================

country_summary = (
    df.groupby("country")
      .agg(
          Sessions=("fullVisitorId", "count"),
          Users=("fullVisitorId", "nunique"),
          Revenue=("transactionRevenue",
                   lambda x: x.fillna(0).sum()/1_000_000)
      )
      .reset_index()
)

country_summary["Revenue"] = country_summary["Revenue"].round(2)

country_summary = (
    country_summary
    .sort_values("Sessions", ascending=False)
)

country_summary.head(10)

,country,Sessions,Users,Revenue
182,United States,39511,33643,168998.63
82,India,6089,5744,141.14
181,United Kingdom,4203,3888,106.31
32,Canada,2747,2404,5187.6
187,Vietnam,2639,2604,0.0
24,Brazil,2507,2395,196.17
171,Thailand,2376,2329,110.96
175,Turkey,2373,2316,0.0
90,Japan,2140,1910,149.71
66,Germany,2051,1919,37.98


In [40]:
top10_country = country_summary.head(10)

top10_country

,country,Sessions,Users,Revenue
182,United States,39511,33643,168998.63
82,India,6089,5744,141.14
181,United Kingdom,4203,3888,106.31
32,Canada,2747,2404,5187.6
187,Vietnam,2639,2604,0.0
24,Brazil,2507,2395,196.17
171,Thailand,2376,2329,110.96
175,Turkey,2373,2316,0.0
90,Japan,2140,1910,149.71
66,Germany,2051,1919,37.98


In [44]:
# =====================================================
# Top 10 Countries by Website Sessions
# Executive Dashboard Style
# =====================================================

import plotly.graph_objects as go

# -----------------------------------------------------
# Prepare Data
# -----------------------------------------------------

top10_country = (
    country_summary
    .sort_values("Sessions", ascending=False)
    .head(10)
    .reset_index(drop=True)
)

top10_country["Share"] = (
    top10_country["Sessions"] /
    top10_country["Sessions"].sum() * 100
).round(1)

top10_country["Rank"] = top10_country.index + 1

top10_country["Label"] = (
    top10_country["Sessions"]
        .map("{:,.0f}".format)
    + " ("
    + top10_country["Share"].astype(str)
    + "%)"
)

# -----------------------------------------------------
# Google Material Colors
# -----------------------------------------------------

colors = [
    "#1967D2",
    "#8AB4F8",
    "#8AB4F8",
    "#D2E3FC",
    "#DADCE0",
    "#DADCE0",
    "#DADCE0",
    "#DADCE0",
    "#DADCE0",
    "#DADCE0"
]

# -----------------------------------------------------
# Figure
# -----------------------------------------------------

fig = go.Figure()

fig.add_trace(

    go.Bar(

        x=top10_country["Sessions"],

        y=top10_country["country"],

        orientation="h",

        marker=dict(
            color=colors,
            line=dict(color="white", width=1)
        ),

        text=top10_country["Label"],

        textposition="auto",

        textfont=dict(
            family="Arial",
            size=13
        ),

        width=0.72,

        cliponaxis=False,

        customdata=top10_country[
            ["Users","Revenue","Share","Rank"]
        ],

        hovertemplate=
        "<b>%{y}</b><br><br>"
        "Sessions : %{x:,}<br>"
        "Users : %{customdata[0]:,}<br>"
        "Revenue : $%{customdata[1]:,.2f}M<br>"
        "Traffic Share : %{customdata[2]}%<br>"
        "Rank : #%{customdata[3]}"
        "<extra></extra>"

    )

)

fig.update_layout(

    template="simple_white",

    title=dict(
        text="<b>Top 10 Markets by Website Sessions</b><br>"
             "<sup>Google Analytics Sample Dataset • Geographic Traffic Distribution</sup>",
        x=0.5,
        font=dict(size=28)
    ),

    height=650,

    margin=dict(
        l=150,
        r=60,
        t=120,
        b=40
    ),

    showlegend=False,

    font=dict(
        family="Arial",
        size=14,
        color="#202124"
    ),

    plot_bgcolor="white",
    paper_bgcolor="white",

    hoverlabel=dict(
        bgcolor="white",
        bordercolor="white",
        font_size=13
    )

)

fig.update_xaxes(

    title="",

    tickformat=",.0f",

    gridcolor="#F1F3F4",

    zeroline=False

)

fig.update_yaxes(

    title="",

    categoryorder="total ascending"

)

fig.show()

In [45]:
# =====================================================
# Global Website Traffic Distribution
# =====================================================

import plotly.express as px

fig = px.choropleth(

    country_summary,

    locations="country",

    locationmode="country names",

    color="Sessions",

    hover_name="country",

    color_continuous_scale=[
        "#E8F0FE",
        "#C6DAFC",
        "#8AB4F8",
        "#4285F4",
        "#1967D2"
    ],

    custom_data=[
        "Users",
        "Revenue",
        "Sessions"
    ]

)

fig.update_traces(

    hovertemplate=

    "<b>%{hovertext}</b><br><br>"

    "Sessions : %{customdata[2]:,.0f}<br>"

    "Users : %{customdata[0]:,.0f}<br>"

    "Revenue : $%{customdata[1]:,.2f}M"

    "<extra></extra>"

)

fig.update_geos(

    showframe=False,

    showcoastlines=False,

    showcountries=True,

    countrycolor="#C4C7C5",

    countrywidth=0.7,

    projection_type="natural earth",

    fitbounds="locations",

    bgcolor="white"

)

fig.update_layout(

    template="simple_white",

    title=dict(

        text="<b>Global Website Traffic Distribution</b><br>"
             "<sup>Google Analytics Sample Dataset • Sessions by Country</sup>",

        x=0.5,

        font=dict(size=28)

    ),

    font=dict(

        family="Arial",

        size=14,

        color="#202124"

    ),

    paper_bgcolor="white",

    plot_bgcolor="white",

    height=700,

    margin=dict(

        l=20,

        r=20,

        t=100,

        b=20

    ),

    coloraxis_showscale=False

)

fig.show()

C:\Users\victus\AppData\Local\Temp\ipykernel_10184\7666853.py:7: DeprecationWarning: The library used by the *country names* `locationmode` option is changing in an upcoming version. Country names in existing plots may not work in the new version. To ensure consistent behavior, consider setting `locationmode` to *ISO-3*.
  fig = px.choropleth(


---
# 📌 Geographic Analysis – Business Insights

## Executive Summary

The geographic analysis reveals that website traffic is concentrated in a few high-performing markets, with the **United States** contributing the largest share of sessions. **India** and the **United Kingdom** are the next strongest traffic sources, indicating significant international reach beyond the primary market.

While traffic is distributed across **191 countries**, user engagement is highly concentrated within the top markets. This suggests that marketing efforts have been particularly successful in a limited number of regions, while many countries represent opportunities for future growth.

---

## 🔍 Key Insights

- 🇺🇸 The **United States** is the primary traffic source and the most valuable market based on website sessions.
- 🌏 Traffic originates from **191 countries**, demonstrating a strong global digital presence.
- 📈 The top markets account for the majority of website sessions, highlighting concentrated audience engagement.
- 🌍 Several countries generate moderate traffic and may represent potential expansion opportunities through localized marketing strategies.
- 💰 Geographic performance should be evaluated alongside conversion rate and revenue to identify high-value markets.

---

## 💼 Business Recommendations

- Prioritize advertising budgets for the highest-performing markets while maintaining campaign efficiency.
- Develop localized campaigns for emerging markets with consistent traffic growth.
- Compare revenue and conversion performance across countries to optimize return on advertising investment (ROAS).
- Monitor regional traffic trends to identify new growth opportunities and changing customer behavior.
- Integrate geographic performance into future budget allocation and forecasting models.

---

## ➡️ Next Analysis

The next section focuses on **Device & Technology Analysis**, where we will examine user behavior across different **device categories**, **browsers**, and **operating systems** to identify optimization opportunities for user experience and campaign performance.

# 5 Device & Technology Analysis📱 

Understanding user devices and technology platforms is essential for optimizing website performance, advertising campaigns, and user experience.

This section evaluates website traffic across device categories, browsers, and operating systems to identify high-performing platforms and opportunities for technical optimization.

In [46]:
# =====================================================
# Device Performance Summary
# =====================================================

device_summary = (
    df.groupby("deviceCategory")
      .agg(
          Sessions=("fullVisitorId", "count"),
          Users=("fullVisitorId", "nunique"),
          Revenue=("transactionRevenue",
                   lambda x: x.fillna(0).sum() / 1_000_000)
      )
      .reset_index()
      .sort_values("Sessions", ascending=False)
)

device_summary

,deviceCategory,Sessions,Users,Revenue
0,desktop,74109,66848,176603.45
1,mobile,22650,20228,4132.0
2,tablet,3241,2913,853.29


In [47]:
device_summary

,deviceCategory,Sessions,Users,Revenue
0,desktop,74109,66848,176603.45
1,mobile,22650,20228,4132.0
2,tablet,3241,2913,853.29


### 📊 Device Performance Overview

The table below summarizes website traffic, unique users, and revenue generated across each device category.

This provides an executive overview of how customers access the website and which devices contribute the greatest business value.

In [48]:
import plotly.express as px

fig = px.bar(

    device_summary,

    x="deviceCategory",

    y="Sessions",

    text="Sessions",

    color="Sessions",

    color_continuous_scale=[
        "#E8F0FE",
        "#8AB4F8",
        "#1967D2"
    ]

)

fig.update_traces(

    texttemplate="%{text:,.0f}",

    textposition="outside"

)

fig.update_layout(

    template="simple_white",

    title=dict(

        text="<b>Website Sessions by Device Category</b><br>"
             "<sup>Google Analytics Sample Dataset</sup>",

        x=0.5,

        font=dict(size=26)

    ),

    height=550,

    coloraxis_showscale=False,

    xaxis_title="",

    yaxis_title="Sessions",

    font=dict(size=14)

)

fig.show()

In [49]:
fig = px.pie(

    device_summary,

    names="deviceCategory",

    values="Sessions",

    hole=0.65,

    color_discrete_sequence=[
        "#1967D2",
        "#8AB4F8",
        "#DADCE0"
    ]

)

fig.update_traces(

    textposition="inside",

    textinfo="percent+label"

)

fig.update_layout(

    template="simple_white",

    title=dict(

        text="<b>Traffic Share by Device Category</b>",

        x=0.5,

        font=dict(size=24)

    ),

    height=550,

    showlegend=False

)

fig.show()

In [50]:
browser_summary = (

    df.groupby("browser")

      .agg(

          Sessions=("fullVisitorId","count"),

          Users=("fullVisitorId","nunique"),

          Revenue=("transactionRevenue",
                   lambda x: x.fillna(0).sum()/1_000_000)

      )

      .reset_index()

      .sort_values("Sessions",ascending=False)

      .head(10)

)

browser_summary

,browser,Sessions,Users,Revenue
6,Chrome,68526,60620,167768.24
23,Safari,20403,18982,4631.33
9,Firefox,4120,3767,7913.68
10,Internet Explorer,2148,2057,176.15
8,Edge,1098,1007,996.51
4,Android Webview,856,811,13.98
21,Opera Mini,712,696,0.0
24,Safari (in-app),691,670,50.93
20,Opera,623,597,37.92
27,UC Browser,264,253,0.0


In [61]:
# =====================================================
# Top 10 Browsers by Website Sessions
# =====================================================

import plotly.graph_objects as go

# -----------------------------
# Prepare Data
# -----------------------------

browser_summary = (
    df.groupby("browser")
      .agg(
          Sessions=("fullVisitorId", "count"),
          Users=("fullVisitorId", "nunique"),
          Revenue=("transactionRevenue",
                   lambda x: x.fillna(0).sum()/1_000_000)
      )
      .reset_index()
      .sort_values("Sessions", ascending=False)
      .head(10)
      .reset_index(drop=True)
)

browser_summary["Revenue"] = browser_summary["Revenue"].round(2)

browser_summary["Share"] = (
    browser_summary["Sessions"] /
    browser_summary["Sessions"].sum() * 100
).round(1)

browser_summary["Rank"] = browser_summary.index + 1

browser_summary["Label"] = (
    browser_summary["Sessions"].map("{:,.0f}".format)
    + " • "
    + browser_summary["Share"].astype(str)
    + "%"
)

# -----------------------------
# Colors
# -----------------------------

colors = ["#174EA6"] + ["#DADCE0"]*(len(browser_summary)-1)

# -----------------------------
# Plot
# -----------------------------

fig = go.Figure()

fig.add_trace(

    go.Bar(

        x=browser_summary["Sessions"],

        y=browser_summary["browser"],

        orientation="h",

        marker=dict(
            color=colors,
            line=dict(color="white", width=1)
        ),

        text=browser_summary["Label"],

        textposition="outside",

        cliponaxis=False,

        width=0.72,

        customdata=browser_summary[
            ["Users","Revenue","Share","Rank"]
        ],

        hovertemplate=
        "<b>%{y}</b><br><br>"
        "Rank : %{customdata[3]}<br>"
        "Sessions : %{x:,}<br>"
        "Users : %{customdata[0]:,}<br>"
        "Revenue : $%{customdata[1]:,.2f} M<br>"
        "Traffic Share : %{customdata[2]}%"
        "<extra></extra>"

    )

)

fig.update_layout(

    template="simple_white",

    title=dict(

        text="<b>Top 10 Browsers by Website Sessions</b><br>"
             "<sup>Google Analytics Sample Dataset • Technology Distribution</sup>",

        x=0.5,

        font=dict(size=26)

    ),

    height=620,

    showlegend=False,

    paper_bgcolor="white",

    plot_bgcolor="white",

    font=dict(
        family="Arial",
        size=14,
        color="#202124"
    ),

    margin=dict(
        l=140,
        r=80,
        t=110,
        b=40
    )

)

fig.update_xaxes(

    title="",

    tickformat=",.0f",

    showgrid=True,

    gridcolor="#ECEFF1",

    zeroline=False

)

fig.update_yaxes(

    title="",

    categoryorder="total ascending"

)

fig.show()

In [55]:
# =====================================================
# Operating System Summary
# =====================================================

os_summary = (
    df.groupby("operatingSystem")
      .agg(
          Sessions=("fullVisitorId", "count"),
          Users=("fullVisitorId", "nunique"),
          Revenue=("transactionRevenue",
                   lambda x: x.fillna(0).sum() / 1_000_000)
      )
      .reset_index()
)

# Round Revenue
os_summary["Revenue"] = os_summary["Revenue"].round(2)

# Calculate Share
os_summary["Share"] = (
    os_summary["Sessions"]
    / os_summary["Sessions"].sum()
    * 100
).round(1)

# Sort and Keep Top 10
os_summary = (
    os_summary
    .sort_values("Sessions", ascending=False)
    .head(10)
    .reset_index(drop=True)
)

os_summary

,operatingSystem,Sessions,Users,Revenue,Share
0,Windows,39310,36099,45290.37,39.3
1,Macintosh,28218,25063,108811.84,28.2
2,Android,13526,12089,3273.61,13.5
3,iOS,11495,10222,1711.68,11.5
4,Linux,3819,3375,4556.79,3.8
5,Chrome OS,2842,2395,17944.45,2.8
6,(not set),548,531,0.0,0.5
7,Windows Phone,136,126,0.0,0.1
8,Samsung,43,42,0.0,0.0
9,BlackBerry,27,26,0.0,0.0


In [62]:
# =====================================================
# Top 10 Operating Systems by Website Sessions
# =====================================================

import plotly.graph_objects as go

# -----------------------------
# Prepare Data
# -----------------------------

os_summary = (
    df.groupby("operatingSystem")
      .agg(
          Sessions=("fullVisitorId", "count"),
          Users=("fullVisitorId", "nunique"),
          Revenue=("transactionRevenue",
                   lambda x: x.fillna(0).sum()/1_000_000)
      )
      .reset_index()
      .sort_values("Sessions", ascending=False)
      .head(10)
      .reset_index(drop=True)
)

os_summary["Revenue"] = os_summary["Revenue"].round(2)

os_summary["Share"] = (
    os_summary["Sessions"] /
    os_summary["Sessions"].sum() * 100
).round(1)

os_summary["Rank"] = os_summary.index + 1

os_summary["Label"] = (
    os_summary["Sessions"].map("{:,.0f}".format)
    + " • "
    + os_summary["Share"].astype(str)
    + "%"
)

# -----------------------------
# Colors
# -----------------------------

colors = ["#174EA6"] + ["#DADCE0"]*(len(os_summary)-1)

# -----------------------------
# Plot
# -----------------------------

fig = go.Figure()

fig.add_trace(

    go.Bar(

        x=os_summary["Sessions"],

        y=os_summary["operatingSystem"],

        orientation="h",

        marker=dict(
            color=colors,
            line=dict(color="white", width=1)
        ),

        text=os_summary["Label"],

        textposition="outside",

        cliponaxis=False,

        width=0.72,

        customdata=os_summary[
            ["Users","Revenue","Share","Rank"]
        ],

        hovertemplate=
        "<b>%{y}</b><br><br>"
        "Rank : %{customdata[3]}<br>"
        "Sessions : %{x:,}<br>"
        "Users : %{customdata[0]:,}<br>"
        "Revenue : $%{customdata[1]:,.2f} M<br>"
        "Traffic Share : %{customdata[2]}%"
        "<extra></extra>"

    )

)

fig.update_layout(

    template="simple_white",

    title=dict(

        text="<b>Top 10 Operating Systems by Website Sessions</b><br>"
             "<sup>Google Analytics Sample Dataset • Technology Distribution</sup>",

        x=0.5,

        font=dict(size=26)

    ),

    height=620,

    showlegend=False,

    paper_bgcolor="white",

    plot_bgcolor="white",

    font=dict(
        family="Arial",
        size=14,
        color="#202124"
    ),

    margin=dict(
        l=170,
        r=80,
        t=110,
        b=40
    )

)

fig.update_xaxes(

    title="",

    tickformat=",.0f",

    showgrid=True,

    gridcolor="#ECEFF1",

    zeroline=False

)

fig.update_yaxes(

    title="",

    categoryorder="total ascending"

)

fig.show()

In [57]:
# =====================================================
# Executive Technology Summary
# =====================================================

technology_summary = pd.DataFrame({

    "Metric":[
        "Most Used Device",
        "Most Used Browser",
        "Most Used Operating System",
        "Total Device Categories",
        "Total Browsers",
        "Total Operating Systems"
    ],

    "Value":[
        device_summary.iloc[0]["deviceCategory"],
        browser_summary.iloc[0]["browser"],
        os_summary.iloc[0]["operatingSystem"],
        df["deviceCategory"].nunique(),
        df["browser"].nunique(),
        df["operatingSystem"].nunique()
    ]

})

technology_summary

,Metric,Value
0,Most Used Device,desktop
1,Most Used Browser,Chrome
2,Most Used Operating System,Windows
3,Total Device Categories,3
4,Total Browsers,29
5,Total Operating Systems,16


### 💡 Operating System Insights

#### Executive Summary

The operating system analysis reveals the technology platforms used by website visitors. A small number of operating systems account for the majority of website traffic, indicating that optimization efforts can be focused on these dominant platforms.

#### Key Findings

- Windows is the most widely used operating system among website visitors.
- Mobile operating systems, including Android and iOS, contribute a significant share of website sessions.
- The top operating systems account for the majority of website traffic, simplifying browser and compatibility testing.
- Cross-platform optimization remains essential for delivering a consistent user experience.

#### Business Recommendations

- Prioritize website optimization and testing for Windows, Android, and iOS.
- Ensure responsive website design across desktop and mobile operating systems.
- Monitor operating system trends regularly to identify changes in user behavior.
- Optimize advertising landing pages for the operating systems generating the highest engagement and conversions.

---

### ➡ Next Section

The next section focuses on **Campaign Performance & Revenue Analysis**, where we evaluate marketing effectiveness, conversion performance, and revenue generated across advertising campaigns.

# 6 Campaign Performance Analysis

Campaign performance analysis evaluates the effectiveness of marketing campaigns in attracting website visitors and generating business value.

This section compares campaign performance using sessions, users, transactions, and revenue to identify high-performing and underperforming campaigns.

In [63]:
# =====================================================
# Campaign Performance Summary
# =====================================================

campaign_summary = (

    df.groupby("campaign")

      .agg(

          Sessions=("fullVisitorId","count"),

          Users=("fullVisitorId","nunique"),

          Transactions=("transactions",
                        lambda x: x.fillna(0).sum()),

          Revenue=("transactionRevenue",
                   lambda x: x.fillna(0).sum()/1_000_000)

      )

      .reset_index()

)

campaign_summary["Revenue"] = campaign_summary["Revenue"].round(2)

campaign_summary = campaign_summary.sort_values(
    "Sessions",
    ascending=False
)

campaign_summary

,campaign,Sessions,Users,Transactions,Revenue
0,(not set),96798,87490,1291,180227.81
5,Data Share Promo,1898,1730,0,0.0
3,AW - Dynamic Search Ads Whole Site,716,628,16,872.23
1,AW - Accessories,575,497,15,488.7
4,AW - Electronics,11,7,0,0.0
2,AW - Apparel,2,2,0,0.0


In [64]:
campaign_kpi = pd.DataFrame({

"Metric":[

"Total Campaigns",

"Top Campaign",

"Total Revenue (Million)",

"Total Transactions"

],

"Value":[

campaign_summary.shape[0],

campaign_summary.iloc[0]["campaign"],

round(campaign_summary["Revenue"].sum(),2),

int(campaign_summary["Transactions"].sum())

]

})

campaign_kpi

,Metric,Value
0,Total Campaigns,6
1,Top Campaign,(not set)
2,Total Revenue (Million),181588.74
3,Total Transactions,1322


In [66]:
# =====================================================
# Campaign Sessions
# =====================================================

import plotly.graph_objects as go

campaign_plot = campaign_summary.sort_values(
    "Sessions",
    ascending=False
).copy()

campaign_plot["Share"] = (
    campaign_plot["Sessions"] /
    campaign_plot["Sessions"].sum() * 100
).round(1)

campaign_plot["Rank"] = campaign_plot.index + 1

campaign_plot["Label"] = campaign_plot["Sessions"].map("{:,.0f}".format)

colors = ["#174EA6"] + ["#DADCE0"]*(len(campaign_plot)-1)

fig = go.Figure()

fig.add_trace(

    go.Bar(

        x=campaign_plot["Sessions"],

        y=campaign_plot["campaign"],

        orientation="h",

        marker=dict(
            color=colors,
            line=dict(color="white", width=1)
        ),

        text=campaign_plot["Label"],

        textposition="outside",

        cliponaxis=False,

        width=0.72,

        customdata=campaign_plot[
            ["Users","Transactions","Revenue","Share"]
        ],

        hovertemplate=
        "<b>%{y}</b><br><br>"
        "Sessions : %{x:,}<br>"
        "Users : %{customdata[0]:,}<br>"
        "Transactions : %{customdata[1]:,.0f}<br>"
        "Revenue : $%{customdata[2]:,.2f} M<br>"
        "Traffic Share : %{customdata[3]}%"
        "<extra></extra>"

    )

)

fig.update_layout(

    template="simple_white",

    title=dict(

        text="<b>Campaign Performance by Website Sessions</b><br>"
             "<sup>Google Analytics Sample Dataset</sup>",

        x=0.5,

        font=dict(size=25)

    ),

    height=500,

    showlegend=False,

    plot_bgcolor="white",

    paper_bgcolor="white"

)

fig.update_xaxes(
    title="",
    tickformat=","
)

fig.update_yaxes(
    title="",
    categoryorder="total ascending"
)

fig.show()

In [67]:
campaign_revenue = campaign_summary.sort_values(
    "Revenue",
    ascending=False
)

fig = go.Figure()

fig.add_trace(

    go.Bar(

        x=campaign_revenue["Revenue"],

        y=campaign_revenue["campaign"],

        orientation="h",

        marker_color="#174EA6",

        text=campaign_revenue["Revenue"].map("${:,.2f}M".format),

        textposition="outside"

    )

)

fig.update_layout(

    template="simple_white",

    title=dict(

        text="<b>Revenue Generated by Campaign</b><br>"
             "<sup>Campaign Performance Comparison</sup>",

        x=0.5,

        font=dict(size=25)

    ),

    height=500,

    showlegend=False

)

fig.update_xaxes(title="Revenue (Million USD)")

fig.update_yaxes(categoryorder="total ascending")

fig.show()

In [68]:
campaign_summary["Conversion Rate"] = (

    campaign_summary["Transactions"] /
    campaign_summary["Sessions"] *100

).round(2)

campaign_summary

,campaign,Sessions,Users,Transactions,Revenue,Conversion Rate
0,(not set),96798,87490,1291,180227.81,1.33
5,Data Share Promo,1898,1730,0,0.0,0.0
3,AW - Dynamic Search Ads Whole Site,716,628,16,872.23,2.23
1,AW - Accessories,575,497,15,488.7,2.61
4,AW - Electronics,11,7,0,0.0,0.0
2,AW - Apparel,2,2,0,0.0,0.0


In [69]:
campaign_conversion = campaign_summary.sort_values(
    "Conversion Rate",
    ascending=False
)

fig = go.Figure()

fig.add_trace(

    go.Bar(

        x=campaign_conversion["Conversion Rate"],

        y=campaign_conversion["campaign"],

        orientation="h",

        marker_color="#34A853",

        text=campaign_conversion["Conversion Rate"].map("{:.2f}%".format),

        textposition="outside"

    )

)

fig.update_layout(

    template="simple_white",

    title=dict(

        text="<b>Campaign Conversion Rate</b><br>"
             "<sup>Marketing Effectiveness Comparison</sup>",

        x=0.5,

        font=dict(size=25)

    ),

    height=500,

    showlegend=False

)

fig.update_xaxes(title="Conversion Rate (%)")

fig.update_yaxes(categoryorder="total ascending")

fig.show()

In [71]:
import numpy as np

In [72]:
ai_campaign = campaign_summary.copy()

conditions = [

    (ai_campaign["Conversion Rate"] >= 2),

    (ai_campaign["Conversion Rate"] >= 1),

    (ai_campaign["Conversion Rate"] < 1)

]

recommendations = [

    "Increase Budget",

    "Monitor Performance",

    "Optimize or Pause"

]

ai_campaign["AI Recommendation"] = np.select(

    conditions,

    recommendations,

    default="Review Campaign"

)

ai_campaign[[
    "campaign",
    "Sessions",
    "Revenue",
    "Conversion Rate",
    "AI Recommendation"
]]

,campaign,Sessions,Revenue,Conversion Rate,AI Recommendation
0,(not set),96798,180227.81,1.33,Monitor Performance
5,Data Share Promo,1898,0.0,0.0,Optimize or Pause
3,AW - Dynamic Search Ads Whole Site,716,872.23,2.23,Increase Budget
1,AW - Accessories,575,488.7,2.61,Increase Budget
4,AW - Electronics,11,0.0,0.0,Optimize or Pause
2,AW - Apparel,2,0.0,0.0,Optimize or Pause


In [73]:
#***************************
campaign_distribution = (
    df["campaign"]
      .value_counts(dropna=False)
      .reset_index()
)

campaign_distribution.columns = ["Campaign", "Sessions"]

campaign_distribution["Percentage"] = (
    campaign_distribution["Sessions"]
    / len(df)
    * 100
).round(2)

campaign_distribution

,Campaign,Sessions,Percentage
0,(not set),96798,96.80
1,Data Share Promo,1898,1.90
2,AW - Dynamic Search Ads Whole Site,716,0.72
3,AW - Accessories,575,0.57
4,AW - Electronics,11,0.01
5,AW - Apparel,2,0.00


In [74]:
pd.crosstab(df["campaign"], df["source"])

source,(direct),(not set),9to5google.com,Partners,adwords.google.com,analytics.google.com,aol,ask,au.search.yahoo.com,aulagrado.esic.edu,...,us.yhs4.search.yahoo.com,vk.com,voice.google.com,wap.sogou.com,web.facebook.com,web.telegram.org,wheretoget.it,yahoo,yahoo.com,youtube.com
campaign,,,,,,,,,,,,,,,,,,,,,
(not set),37669,9,1,1,7,1841,1,30,4,1,...,1,8,1,2,1,1,2,190,1,23973
AW - Accessories,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
AW - Apparel,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
AW - Dynamic Search Ads Whole Site,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
AW - Electronics,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
Data Share Promo,0,0,0,1898,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [75]:
pd.crosstab(df["campaign"], df["medium"])

medium,(none),(not set),affiliate,cpc,cpm,organic,referral
campaign,,,,,,,
(not set),37669,17,0,113,643,29142,29214
AW - Accessories,0,0,0,575,0,0,0
AW - Apparel,0,0,0,2,0,0,0
AW - Dynamic Search Ads Whole Site,0,0,0,716,0,0,0
AW - Electronics,0,0,0,11,0,0,0
Data Share Promo,0,0,1898,0,0,0,0


In [76]:
df.groupby("channelGrouping").size().sort_values(ascending=False)

channelGrouping
Organic Search    42337
Social            25259
Direct            16035
Referral          11335
Paid Search        2471
Affiliates         1898
Display             648
(Other)              17
dtype: int64